### PR TITLE
[DataGrid] Filtering performance: use `Set()` when applicable

### PR DESCRIPTION
**Pull request by romgrk**
Last part of #9120

*edit: Read the summary below, these results are outdated.*

This PR improves performance by using `Set` where applicable.

|  | Elapsed (ms) |
| --- | --- |
| Without `Set` | `56.1 (+-2.5)` |
| With `Set` | `35.6 (+-1.8)` |

Test: 100,000 rows of the commodity dataset, filter one string column with the string "am". We observe gains of about 35% for this test.

Note that I was thinking we'd need to update the selectors to keep returning `objects` instead of `Sets`, but I haven't found any public selector that exposes them, they're all `@ignore Do not document`. If you see anything that would break semver guarantees, please point it out.

### CODE REVIEW

File: scripts/x-data-grid-premium.exports.json [hunk 1]
```diff
@@ -574,7 +575,6 @@
   { "name": "GridVisibilityOffIcon", "kind": "Variable" },
   { "name": "gridVisibleColumnDefinitionsSelector", "kind": "Variable" },
   { "name": "gridVisibleColumnFieldsSelector", "kind": "Variable" },
-  { "name": "gridVisibleRowsLookupSelector", "kind": "Variable" },
```
Comment (cherniavskii) [inline scripts/x-data-grid-premium.exports.json:577]: This feels risky TBH. While `gridVisibleRowsLookupSelector` is not documented, it's exported and not marked as unstable/experimental.
Should we wait with this PR until v7 alpha?

File: packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGroupingPreProcessors.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ import {
   createRowTree,
   updateRowTree,
   RowTreeBuilderGroupingCriterion,
-  getVisibleRowsLookup,
+  getHiddenRowsLookup,
```
Comment (cherniavskii) [inline packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGroupingPreProcessors.ts:20]: > However, making that change lost nearly all of the performance improvements observed in the first step.

I don't understand what caused the loss of the performance improvements.
First, you tried using `Set` to keep all the visible rows.
Then, you reversed it to only keep the hidden rows. Did this increase the number of records in the `Set` drastically?
Or is it something else that impacted performance?

Comment (romgrk) [reply packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGroupingPreProcessors.ts:20]: Also didn't make sense to me, but I haven't had enough time to investigate further.

**Comment by romgrk** [comment]
Although the performance gains are interesting I don't have the bandwidth to continue exploring this one so I'll close it for now. If we look into it again, I'm pretty sure we can expose a backwards compatible interface even if we change the internal model.

**Pull request closed without merging**